### PR TITLE
[MLIR][TORCH] Add E2E support for torch.arange op

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
This commit adds lowering of `aten.arange.start_step` op.
This commit decomposes `aten.arange` and `aten.arange.start` into
`aten.arange.start_step` op.

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

### CODE REVIEW

**Review by pashu123 — CHANGES_REQUESTED** [review]

**Review by gprateek93 — COMMENTED** [review]
Nice work. One minor comment.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -563,6 +563,51 @@ class DecomposeAtenEmptyLikeOp : public OpRewritePattern<AtenEmptyLikeOp> {
...
+    Location loc = op.getLoc();
+    Value start, step;
+    Value end = op.end();
+    Value result = op.result();
+    BaseTensorType resultType = result.getType().cast<BaseTensorType>();
+    bool isFloatDtype =
+        resultType.hasDtype() && resultType.getDtype().isa<mlir::FloatType>();
+    // If the op is a AtenArangeStartOp then it has a start value. Otherwise we
+    // set it as default value 0.
+    if (isa<AtenArangeStartOp>(op)) {
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:595]: I'm always a bit unsure about combining templates and dynamic casts. The main advantage of a template is that it allows compile time polymorphism (which can have a speed benefit), but the dynamic cast counteracts the benefit by requiring type information during runtime. 

One way of avoiding this issue is to have the template be a helper function that takes as arguments the `op` and the `start` value, then the two classes for decomposing `aten.arange` and `aten.arange.start` can call this function. In other words, we are abstracting too early. What the two ops have in common is not the decomposition, but the decomposition once the `start` value has been determined

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Thanks for this! A few comments below

**Review by ramiro050 — APPROVED** [review]
Thanks for this. LGTM!

**Pull request merged**